feat(payments): add storybook for UI work

lmorchard opened the pull request.

ianb commented on packages/fxa-payments-server/src/components/LoadingOverlay.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+  .add('isLoading = true', () => (
+    <LoadingOverlay isLoading={true} />
+  ))
+  .add('isLoading = false', () => (
```
This shows up as... a blank page? Maybe that makes sense, but also seems weird...?

lmorchard replied:
The loader shouldn't appear at all when not loading, so technically that's what this shows.

ianb commented on packages/fxa-payments-server/src/components/ProductValueProposition/index.stories.tsx:
```diff
@@ -0,0 +1,37 @@
+    <MockLoader>
+      <ProductValueProposition plan={mockPlan} />
+    </MockLoader>
+  ))
```
The default one gives me this, is that expected?

![image](https://user-images.githubusercontent.com/44390/58832537-c4c19680-8614-11e9-8bd5-eee696301c55.png)

lmorchard replied:
Yeah, this is just demo content right now - namely a [JSON.stringify() of the plan data](https://github.com/mozilla/fxa/blob/master/packages/fxa-payments-server/src/components/ProductValueProposition/index.tsx#L14).

ianb commented on packages/fxa-payments-server/src/components/Profile.stories.tsx:
```diff
@@ -0,0 +1,16 @@
+import { Profile } from './Profile';
+
+storiesOf('components/Profile', module)
+  .add('with all profile data', () => (
```
This looks pretty unwrapped:

![image](https://user-images.githubusercontent.com/44390/58832602-e7ec4600-8614-11e9-97dc-1a85f0958733.png)

For Storybook, should we make sure it's inside a representative container?

lmorchard replied:
Sure, though we don't have one implemented yet.

ianb commented on packages/fxa-payments-server/src/routes/Product.tsx:
```diff
@@ -15,7 +15,7 @@ import {
 import { useCheckboxState } from '../lib/hooks';
 
-type ProductProps = {
+export type ProductProps = {
```
I'm guessing this indicates a bug in the mock setup?

![image](https://user-images.githubusercontent.com/44390/58832750-40234800-8615-11e9-8582-00b2829e1c6e.png)

lmorchard replied:
This will be a slight bitrot from other PRs getting merged before this one

ianb commented on packages/fxa-payments-server/src/routes/Subscriptions/index.stories.tsx:
```diff
@@ -0,0 +1,109 @@
+        }} />
+      </MockApp>
+    ))
+    .add('error', () => (
```
I suppose `[object Object]` here isn't a bug in the Storybook, just in code that Storybook is calling...

![image](https://user-images.githubusercontent.com/44390/58832819-65b05180-8615-11e9-8654-7ebd182b7b14.png)

lmorchard replied:
Yeah, this is just what the component does right now. Planning to change this when implementing error states for this component

ianb commented on packages/fxa-payments-server/package.json:
```diff
@@ -11,7 +11,9 @@
-    "test:ts": "cross-env TS_NODE_COMPILER_OPTIONS='{\"module\":\"commonjs\", \"jsx\": \"react\"}' nyc --reporter=html --reporter=text mocha -r ts-node/register -r @babel/register -r jsdom-global/register -r test-setup.ts src/**/*-test.{ts,tsx} --recursive --timeout 5000"
+    "test:ts": "cross-env TS_NODE_COMPILER_OPTIONS='{\"module\":\"commonjs\", \"jsx\": \"react\"}' nyc --reporter=html --reporter=text mocha -r ts-node/register -r @babel/register -r jsdom-global/register -r test-setup.ts src/**/*-test.{ts,tsx} --recursive --timeout 5000",
+    "storybook": "start-storybook -p 6006",
+    "build-storybook": "build-storybook"
```
I wasn't able to run `npm run build-storybook` due to some TypeScript errors, like:

```
ERR! => Failed to build the preview                                           0 ERR! /Users/ianbicking/src/fxa/packages/fxa-payments-server/src/components/Profile.stories.tsx
ERR! TypeScript error in /Users/ianbicking/src/fxa/packages/fxa-payments-server/src/components/Profile.stories.tsx(10,7):
ERR! Type '{ avatar: string; displayName: string; email: string; }' is missing the following properties from type 'Profile': amrValues, avatarDefault, locale, twoFactorAuthentication, uid  TS2739
ERR! 
ERR!      8 |       error: null,
ERR!      9 |       loading: false,
ERR!   > 10 |       result: {
ERR!        |       ^
ERR!     11 |         avatar: '',
ERR!     12 |         displayName: 'Foo Barson',
ERR!     13 |         email: 'foo@example.com',
ERR! /Users/ianbicking/src/fxa/packages/fxa-payments-server/src/routes/Product.stories.tsx
ERR! TypeScript error in /Users/ianbicking/src/fxa/packages/fxa-payments-server/src/routes/Product.stories.tsx(45,5):
```

lmorchard replied:
D'oh - looks like some code that I split the wrong way when I untangled that PR from earlier. Will update with fixes

lmorchard commented on packages/fxa-payments-server/src/components/ProductValueProposition/index.tsx:
```diff
@@ -25,7 +25,7 @@ const ProductValueProposition = ({
   return (
     <div className="productDetails">
-      <p>For {currency} {amount} per {interval}, your {plan_name} includes:</p>
+      <p>For {currency} {amount / 100.0} per {interval}, your {plan_name} includes:</p>
```
Yeah, doesn't really belong in this PR. But, [amount in cents is what Stripe offers](https://stripe.com/docs/api/plans/object#plan_object-amount) and subhub passes along.

lmorchard commented on packages/fxa-payments-server/.storybook/config.js:
```diff
@@ -0,0 +1,16 @@
+  [
+    reqFromComponents,
+    reqFromRoutes
+  ].forEach(req => req.keys().forEach(filename => req(filename)));
```
This is a usage of [webpack require context API](https://webpack.js.org/guides/dependency-management#context-module-api) - `req` is a require context. This lets us specify dependencies by regex/glob at build time and then dynamically import them at run time.

lmorchard commented on packages/fxa-payments-server/src/components/DialogMessage.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+storiesOf('components/DialogMessage', module)
+  .add('error', () => (
+    <DialogMessage className="error" onDismiss={action('dismiss')}>
+      This is a dialog message
```
Sure, haven't gotten to actually implementing stuff from the mockups yet though. Just trying to lay down a foundation of starter stories here

lmorchard commented:
Rebased against master, fixed type errors and a build error in the payment form. Other stuff around what the components are actually doing when rendered should be addressed as part of other issues related to visuals & UX

lmorchard commented:
Oh boy do I love `package-lock.json` in a git repository. Rebased against master, I think things are still good. Both `npm run storybook` and `npm run build-storybook` still work. Going to wait to see if circleci goes green, then merge.
